File: fast_sync_dart_client/fast_sync_client/lib/src/implementation/metadata/syncable_metadata.dart
import 'package:fast_sync_client/src/absraction/constants.dart';
import 'package:fast_sync_client/src/absraction/metadata/isync_metadata.dart';
import 'package:fast_sync_client/src/absraction/models/sync_zone_restriction.dart';

class SyncMetadata implements ISyncMetadata {
  @override
  String id;

  @override
  String? syncZone;

  @override
  final String type;

  @override
  num version;

  @override
  num timestamp;

  @override
  int? syncOperation;

  SyncMetadata(
      {required this.id,
      required this.version,
      required this.timestamp,
      required this.type});

  @override
  String? getSyncZone() {
    return syncZone;
  }

  @override
  String computeSyncZone(SyncZoneRestrictionEnum syncZoneRestrictionType) {
    if (syncZoneRestrictionType == SyncZoneRestrictionEnum.restricted) {
      return syncZone!;
    } else {
      return Constants.globalSyncZoneRestriction;
    }
  }
}

File: fast-sync-typescript-client-sdk/packages/fast-sync-client/src/implementation/metadata/syncable_metadata.ts
import { Constants } from "../../abstraction/constants";
import { ISyncMetadata } from "../../abstraction/metadata/isync_metadata";
import { SyncOperationEnum } from "../../abstraction/metadata/isync_operation";
import { v4 as uuid } from "uuid";
import { SyncZoneRestrictionEnum } from "../../abstraction/models/sync_zone_restriction";

export class SyncMetadata implements ISyncMetadata {
  id: string;
  timestamp: number;
  syncOperation: SyncOperationEnum;

  static create(metadata: SyncMetadata) {
    let syncOperationMetadata = new SyncMetadata(
      metadata.type,
      metadata.version,
      metadata.syncZone
    );
    return syncOperationMetadata;
  }
  constructor(
    public type: string,
    public version: number,
    public syncZone: string
  ) {
    this.id = uuid();
    this.timestamp = new Date().getTime() / 1000;
  }

  getSyncZone() {
    return this.syncZone;
  }

  computeSyncZone(syncZoneRestrictionType: SyncZoneRestrictionEnum) {
    if (syncZoneRestrictionType == SyncZoneRestrictionEnum.restricted)
      return this.syncZone;
    else return Constants.globalSyncZoneRestriction;
  }
}
